File: leetcode/python/49_group_anagrams.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anagrams = defaultdict(list[int])
        for word in strs:
            letters = [0] * 26
            for w in word:
                letters[ord(w) - ord('a')] += 1
            key = tuple(letters)
            anagrams[key].append(word)
        return list(anagrams.values())

s = Solution()
print(s.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
print(s.groupAnagrams([""]))
print(s.groupAnagrams(["a"]))

File: leetcode/rust/src/49_group_anagrams.rs
impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map = std::collections::HashMap::new();
        for s in strs {
            let mut key = s.chars().collect::<Vec<char>>();
            key.sort();
            map.entry(key).or_insert(Vec::new()).push(s);
        }

        map.into_iter().map(|(_, v)| v).collect()
    }
}